tests/test_bounce_integral.py

```diff
@@ -0,0 +1,703 @@
+    #       Response: Currently the API is such that the method does all the
+    #                 above preprocessing for you. Let's test it for correctness
+    #                 first then do this later.
+    bi, items = bounce_integral(
+        eq,
+        rho=np.unique(coords1[:, 0]),
+        alpha=alpha,
+        zeta=zeta,
+        return_items=True,
+        check=True,
+    )
+    grid = items["grid"]
+    grid._unique_zeta_idx = np.unique(grid.nodes[:, 2], return_index=True)[1]
+
+    data_keys = [
+        "|grad(psi)|^2",
+        "grad(psi)",
+        "B",
+        "iota",
+        "|B|",
+        "B^zeta",
+        "cvdrift0",
+        "cvdrift",
+        "gbdrift",
+    ]
+
+    data = eq.compute(data_keys, grid=grid, data=items["data"])
+
+    psib = data_eq["psi"][-1]
+
+    # signs
+    sign_psi = psib / np.abs(psib)
+    sign_iota = iotas / np.abs(iotas)
+
+    # normalizations
+    Lref = data_eq["a"]
+    Bref = 2 * np.abs(psib) / Lref**2
+
+    modB = data["|B|"]
+    bmag = modB / Bref
+
+    x = Lref * np.sqrt(psi)
+    s_hat = -x / iotas * shears / Lref
+
+    iota = data["iota"]
+    gradpar = Lref * data["B^zeta"] / modB
+
+    ## Comparing coefficient calculation here with coefficients from compute/_mtric
+    cvdrift = -2 * sign_psi * Bref * Lref**2 * np.sqrt(psi) * data["cvdrift"]
+    gbdrift = -2 * sign_psi * Bref * Lref**2 * np.sqrt(psi) * data["gbdrift"]
+
+    a0_over_R0 = Lref * np.sqrt(psi)
+
+    bmag_an = np.mean(bmag) * (1 - a0_over_R0 * np.cos(theta_PEST))
+    np.testing.assert_allclose(bmag, bmag_an, atol=5e-3, rtol=5e-3)
+
+    gradpar_an = 2 * Lref * iota * (1 - a0_over_R0 * np.cos(theta_PEST))
+    np.testing.assert_allclose(gradpar, gradpar_an, atol=9e-3, rtol=5e-3)
+
+    dPdrho = np.mean(-0.5 * (cvdrift - gbdrift) * modB**2)
+    alpha_MHD = -dPdrho * 1 / iota**2 * 0.5
+
+    grad_psi = data["grad(psi)"]
+    grad_alpha = data["grad(alpha)"]
+
+    gds21 = -sign_iota * np.array(dot(grad_psi, grad_alpha)) * s_hat / Bref
+
+    fudge_factor2 = 0.19
+    gbdrift_an = fudge_factor2 * (
+        -1 * s_hat + (np.cos(theta_PEST) - 1.0 * gds21 / s_hat * np.sin(theta_PEST))
+    )
+
+    fudge_factor3 = 0.07
+    cvdrift_an = gbdrift_an + fudge_factor3 * alpha_MHD / bmag**2
+
+    # Comparing coefficients with their analytical expressions
+    np.testing.assert_allclose(gbdrift, gbdrift_an, atol=1.5e-2, rtol=5e-3)
+    np.testing.assert_allclose(cvdrift, cvdrift_an, atol=9e-3, rtol=5e-3)
+
+    # Values of pitch angle for which to evaluate the bounce averages
+    pitch_res = 11
+    pitch = np.linspace(1 / np.max(bmag), 1 / np.min(bmag), pitch_res).reshape(
+        pitch_res, -1
+    )
+
+    bavg_drift_an = (
+        0.5 * cvdrift_an * ellipe(pitch)
+        + gbdrift_an * ellipk(pitch)
+        + dPdrho / bmag**2 * ellipe(pitch)
+    )
+
+    # The quantities are already calculated along a field line
+    bavg_drift_num = bi(
```

Comment:
This integral is becoming nan due to the presence of 1/sqrt(1-lambda B). I think we should only include sqrt(1-lambda B) terms when we are applying the quadrature.

Reply:
Is every term nan or just some? It may be normal for it to be sparse and we just need to filter out the nan values

Reply:
I get an error due to invalid value in sqrt() occurring from the argument to the function, so the argument given the bi function has a negative in the sqrt. So  the supplied pitch values are an issue.

Reply:
I am only checking after bi. After bi everything is nan. Yes, before bounce_averaging, for a given lambda there will be bmag values where lambda*bmag > 1 which is why I am saying that maybe the sqrt(1-lambda B) and 1/sqrt(1-lambda B) must be multiplied during the quadrature.